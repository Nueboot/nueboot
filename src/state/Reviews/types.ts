import { BootReview } from 'api/types';

export interface ReviewsState {
  readonly byId: { [key: string]: BootReview[] };
}
