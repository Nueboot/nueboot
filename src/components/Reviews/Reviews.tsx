import React from 'react';
import { Button, Heading } from '../Styled';

export interface ReviewsProps {
  id: string;
}

const Reviews: React.SFC<ReviewsProps> = () => {
  return (
    <section className="Reviews">
      <Heading size={2}>Reviews</Heading>
      <Button>New Review</Button>
    </section>
  );
};

export default Reviews;
