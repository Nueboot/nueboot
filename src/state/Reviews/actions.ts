import { addReview, fetchBootReview } from 'api';
import { BootReview } from 'api/types';

export const createReview = (review: BootReview, user: string) => dispatch => {
  addReview(review, user).then(dispatch(fetchReview(review.id)));
};

export const receiveReview = (reviews: BootReview[]) => ({
  type: 'REVIEWS.RECEIVE_REVIEW',
  payload: {
    ...reviews,
  },
});

export const fetchReview = (id: string) => dispatch => {
  fetchBootReview(id).then(reviews => {
    dispatch(receiveReview([reviews as BootReview]));
  });
};
