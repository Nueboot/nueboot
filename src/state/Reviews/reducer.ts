const ReviewsReducer = (state = {}, action) => {
  switch (action.type) {
    case 'REVIEWS.RECEIVE_REVIEW':
      return {
        ...action.payload,
      };
    default:
      return state;
  }
};

export default ReviewsReducer;
