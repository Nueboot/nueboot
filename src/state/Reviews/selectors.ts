import { State } from 'state/types';

export const getReviewsForBoot = (state: State, id: string) => {
  const reviews = state.reviews[id];
  if (reviews) {
    return Object.keys(reviews).map(key => {
      return { id: key, ...reviews[key] };
    });
  }
  return [];
};
