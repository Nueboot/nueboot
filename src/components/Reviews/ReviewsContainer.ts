import { BootReview } from 'api/types';
import { connect } from 'react-redux';
import { createReview, fetchReview } from 'state/Reviews/actions';
import { getReviewsForBoot } from 'state/Reviews/selectors';
import { getCurrentUserName } from 'state/Session/selectors';
import Reviews from './Reviews';

const mapStatetoProps = (state, ownProps) => ({
  reviews: getReviewsForBoot(state, ownProps.id),
  currentUser: getCurrentUserName(state),
});

const mapDispatchToProps = dispatch => ({
  submitReview: (review: BootReview, user) =>
    dispatch(createReview(review, user)),
  getReviews: (id: string) => dispatch(fetchReview(id)),
});

export default connect(
  mapStatetoProps,
  mapDispatchToProps
)(Reviews);
