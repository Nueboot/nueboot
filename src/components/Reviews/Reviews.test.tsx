import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import Reviews, { ReviewsProps } from './Reviews';

describe('Reviews', () => {
  let component: ShallowWrapper<ReviewsProps>;

  beforeEach(() => {
    const props: ReviewsProps = {
      reviews: [],
      currentUser: '',
      id: '1',
      submitReview: jest.fn(),
      getReviews: jest.fn(),
    };
    component = shallow(<Reviews {...props} />);
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
