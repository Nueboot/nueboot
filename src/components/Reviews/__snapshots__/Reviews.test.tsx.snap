// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews renders correctly 1`] = `
<section
  className="reviews"
>
  <div
    className="flex justify-between"
  >
    <styled.h1
      className="ttu dib"
      size={2}
    >
      Reviews
    </styled.h1>
    <styled.button
      className="bg-green white"
    >
      New Review
    </styled.button>
  </div>
</section>
`;
